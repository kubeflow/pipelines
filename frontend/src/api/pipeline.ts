import { Parameter } from './parameter';

export class Pipeline {
  public id: number;
  public created_at: string;
  public name: string;
  public description?: string;
  public package_id: number;
  public schedule: string;
  public enabled: boolean;
  public enabled_at: string;
  public parameters?: Parameter[];

  constructor() {
    this.created_at = new Date().toISOString();
    this.name = '';
    this.description = '';
    this.package_id = -1;
    this.schedule = '';
    this.enabled = false;
    this.enabled_at = new Date().toISOString();
    this.parameters = [];
  }
}
