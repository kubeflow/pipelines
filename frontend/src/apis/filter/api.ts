/// <reference path="./custom.d.ts" />
// tslint:disable
/**
 * backend/api/filter.proto
 * No description provided (generated by Swagger Codegen https://github.com/swagger-api/swagger-codegen)
 *
 * OpenAPI spec version: version not set
 * 
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */


import * as portableFetch from "portable-fetch";
import { Configuration } from "./configuration";

const BASE_PATH = "http://localhost".replace(/\/+$/, "");

/**
 *
 * @export
 */
export const COLLECTION_FORMATS = {
    csv: ",",
    ssv: " ",
    tsv: "\t",
    pipes: "|",
};

/**
 *
 * @export
 * @interface FetchAPI
 */
export interface FetchAPI {
    (url: string, init?: any): Promise<Response>;
}

/**
 *  
 * @export
 * @interface FetchArgs
 */
export interface FetchArgs {
    url: string;
    options: any;
}

/**
 * 
 * @export
 * @class BaseAPI
 */
export class BaseAPI {
    protected configuration: Configuration;

    constructor(configuration?: Configuration, protected basePath: string = BASE_PATH, protected fetch: FetchAPI = portableFetch) {
        if (configuration) {
            this.configuration = configuration;
            this.basePath = configuration.basePath || this.basePath;
        }
    }
};

/**
 * 
 * @export
 * @class RequiredError
 * @extends {Error}
 */
export class RequiredError extends Error {
    name: "RequiredError"
    constructor(public field: string, msg?: string) {
        super(msg);
    }
}

/**
 * Filter is used to filter resources returned from a ListXXX request.  Example filters: 1) Filter runs with status = 'Running' filter {   predicate {     key: \"status\"     op: EQUALS     string_value: \"Running\"   } }  2) Filter runs that succeeded since Dec 1, 2018 filter {   predicate {     key: \"status\"     op: EQUALS     string_value: \"Succeeded\"   }   predicate {     key: \"created_at\"     op: GREATER_THAN     timestamp_value {       seconds: 1543651200     }   } }  3) Filter runs with one of labels 'label_1' or 'label_2'  filter {   predicate {     key: \"label\"     op: IN     string_values {       value: 'label_1'       value: 'label_2'     }   } }
 * @export
 * @interface ApiFilter
 */
export interface ApiFilter {
    /**
     * All predicates are AND-ed when this filter is applied.
     * @type {Array&lt;ApiPredicate&gt;}
     * @memberof ApiFilter
     */
    predicates?: Array<ApiPredicate>;
}

/**
 * 
 * @export
 * @interface ApiIntValues
 */
export interface ApiIntValues {
    /**
     * 
     * @type {Array&lt;number&gt;}
     * @memberof ApiIntValues
     */
    values?: Array<number>;
}

/**
 * 
 * @export
 * @interface ApiLongValues
 */
export interface ApiLongValues {
    /**
     * 
     * @type {Array&lt;string&gt;}
     * @memberof ApiLongValues
     */
    values?: Array<string>;
}

/**
 * Predicate captures individual conditions that must be true for a resource being filtered.
 * @export
 * @interface ApiPredicate
 */
export interface ApiPredicate {
    /**
     * 
     * @type {PredicateOp}
     * @memberof ApiPredicate
     */
    op?: PredicateOp;
    /**
     * 
     * @type {string}
     * @memberof ApiPredicate
     */
    key?: string;
    /**
     * 
     * @type {number}
     * @memberof ApiPredicate
     */
    int_value?: number;
    /**
     * 
     * @type {string}
     * @memberof ApiPredicate
     */
    long_value?: string;
    /**
     * 
     * @type {string}
     * @memberof ApiPredicate
     */
    string_value?: string;
    /**
     * Timestamp values will be converted to Unix time (seconds since the epoch) prior to being used in a filtering operation.
     * @type {Date}
     * @memberof ApiPredicate
     */
    timestamp_value?: Date;
    /**
     * Array values below are only meant to be used by the IN operator.
     * @type {ApiIntValues}
     * @memberof ApiPredicate
     */
    int_values?: ApiIntValues;
    /**
     * 
     * @type {ApiLongValues}
     * @memberof ApiPredicate
     */
    long_values?: ApiLongValues;
    /**
     * 
     * @type {ApiStringValues}
     * @memberof ApiPredicate
     */
    string_values?: ApiStringValues;
}

/**
 * 
 * @export
 * @interface ApiStringValues
 */
export interface ApiStringValues {
    /**
     * 
     * @type {Array&lt;string&gt;}
     * @memberof ApiStringValues
     */
    values?: Array<string>;
}

/**
 * Op is the operation to apply.   - EQUALS: Operators on scalar values. Only applies to one of |int_value|, |long_value|, |string_value| or |timestamp_value|.  - IN: Checks if the value is a member of a given array, which should be one of |int_values|, |long_values| or |string_values|.  - IS_SUBSTRING: Checks if the value contains |string_value| as a substring match. Only applies to |string_value|.
 * @export
 * @enum {string}
 */
export enum PredicateOp {
    UNKNOWN = <any> 'UNKNOWN',
    EQUALS = <any> 'EQUALS',
    NOTEQUALS = <any> 'NOT_EQUALS',
    GREATERTHAN = <any> 'GREATER_THAN',
    GREATERTHANEQUALS = <any> 'GREATER_THAN_EQUALS',
    LESSTHAN = <any> 'LESS_THAN',
    LESSTHANEQUALS = <any> 'LESS_THAN_EQUALS',
    IN = <any> 'IN',
    ISSUBSTRING = <any> 'IS_SUBSTRING'
}


