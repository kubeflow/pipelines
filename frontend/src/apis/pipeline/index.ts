// tslint:disable
/**
 * backend/api/pipeline.proto
 * No description provided (generated by Swagger Codegen https://github.com/swagger-api/swagger-codegen)
 *
 * OpenAPI spec version: version not set
 *
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */

export * from './api';
export * from './configuration';
