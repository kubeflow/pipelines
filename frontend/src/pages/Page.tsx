/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { RouteComponentProps } from 'react-router';
import { ToolbarProps } from '../components/Toolbar';
import { BannerProps } from '../components/Banner';
import { SnackbarProps } from '@material-ui/core/Snackbar';
import { DialogProps } from '../components/Router';
import { errorToMessage } from '../lib/Utils';

export interface PageProps extends RouteComponentProps {
  toolbarProps: ToolbarProps;
  updateBanner: (bannerProps: BannerProps) => void;
  updateDialog: (dialogProps: DialogProps) => void;
  updateSnackbar: (snackbarProps: SnackbarProps) => void;
  updateToolbar: (toolbarProps: Partial<ToolbarProps>) => void;
}

export abstract class Page<P, S> extends React.Component<P & PageProps, S> {
  protected _isMounted = true;

  constructor(props: any) {
    super(props);
    this.props.updateToolbar(this.getInitialToolbarState());
  }

  public abstract render(): JSX.Element;

  public abstract getInitialToolbarState(): ToolbarProps;

  public abstract refresh(): Promise<void>;

  public componentWillUnmount(): void {
    this.clearBanner();
    this._isMounted = false;
  }

  public componentDidMount(): void {
    this.clearBanner();
  }

  public clearBanner(): void {
    this.props.updateBanner({});
  }

  public async showPageError(message: string, error?: Error, mode?: 'error' | 'warning'): Promise<void> {
    const errorMessage = await errorToMessage(error);
    this.props.updateBanner({
      additionalInfo: errorMessage ? errorMessage : undefined,
      message: message + (errorMessage ? ' Click Details for more information.' : ''),
      mode: mode || 'error',
      refresh: this.refresh.bind(this),
    });
  }

  public showErrorDialog(title: string, content: string): void {
    this.props.updateDialog({
      buttons: [{ text: 'Dismiss' }],
      content,
      title,
    });
  }

  protected setStateSafe(newState: Partial<S>, cb?: () => void): void {
    if (this._isMounted) {
      this.setState(newState as any, cb);
    }
  }
}
