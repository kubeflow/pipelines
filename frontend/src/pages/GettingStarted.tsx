/*
 * Copyright 2019 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Markdown from 'markdown-to-jsx';
import * as React from 'react';
import { classes, cssRaw } from 'typestyle';
import { ApiFilter, PredicateOp } from '../apis/filter/api';
import { AutoLink } from '../atoms/ExternalLink';
import { RoutePageFactory } from '../components/Router';
import { ToolbarProps } from '../components/Toolbar';
import SAMPLE_CONFIG from '../config/sample_config_from_backend.json';
import { commonCss, padding } from '../Css';
import { Apis } from '../lib/Apis';
import Buttons from '../lib/Buttons';
import { Page } from './Page';

const DEMO_PIPELINES: string[] = SAMPLE_CONFIG;
const DEMO_PIPELINES_ID_MAP = {
  control: 3,
  data: 2,
  tfx: 1,
  xgboost: 0,
  v2python: 4,
};

const PAGE_CONTENT_MD = ({
  control,
  data,
  tfx,
  xgboost,
  v2python,
}: {
  control: string;
  data: string;
  tfx: string;
  xgboost: string;
  v2python: string;
}) => `
<br/>

## Build your own pipeline with

  * Kubeflow Pipelines [SDK](https://www.kubeflow.org/docs/pipelines/sdk/)
    * Try out [the new Kubeflow Pipelines SDK v2 (Beta)](https://www.kubeflow.org/docs/components/pipelines/sdk/v2/v2-compatibility/)
  * TensorFlow Extended (TFX) [SDK](https://www.tensorflow.org/tfx/guide) with end-to-end ML Pipeline Template ([Open a Cloud AI Platform Notebook on Google Cloud](https://console.cloud.google.com/mlengine/notebooks/deploy-notebook?q=download_url%3Dhttps%253A%252F%252Fraw.githubusercontent.com%252Ftensorflow%252Ftfx%252Fmaster%252Fdocs%252Ftutorials%252Ftfx%252Ftemplate.ipynb))

<br/>

## Demonstrations and Tutorials
This section contains demo and tutorial pipelines.

**Demos** - Try an end-to-end demonstration pipeline.

  * [TFX pipeline demo with Estimator](${tfx}) - Classification pipeline with model analysis, based on a public BigQuery dataset of taxicab trips. [source code](https://github.com/kubeflow/pipelines/tree/master/samples/core/parameterized_tfx_oss)
  * [XGBoost Pipeline demo](${xgboost}) - An example of end-to-end iterative XGBoost model training. [source code](https://github.com/kubeflow/pipelines/tree/master/samples/core/train_until_good)

<br/>

**Tutorials** - Learn pipeline concepts by following a tutorial.

* [V2 Python function based components](${v2python}) - Shows different component input and output options for KFP v2 components. [source code](https://github.com/kubeflow/pipelines/blob/master/samples/v2/lightweight_python_functions_v2_pipeline/lightweight_python_functions_v2_pipeline.py)
* [Data passing in Python components](${data}) - Shows how to pass data between Python components. [source code](https://github.com/kubeflow/pipelines/tree/master/samples/tutorials/Data%20passing%20in%20python%20components)
* [DSL - Control structures](${control}) - Shows how to use conditional execution and exit handlers. [source code](https://github.com/kubeflow/pipelines/tree/master/samples/tutorials/DSL%20-%20Control%20structures)

Want to learn more? [Learn from sample and tutorial pipelines.](https://www.kubeflow.org/docs/pipelines/tutorials/)
`;

cssRaw(`
.kfp-start-page li {
  font-size: 14px;
  margin-block-start: 0.83em;
  margin-block-end: 0.83em;
  margin-left: 2em;
}
.kfp-start-page p {
  font-size: 14px;
  margin-block-start: 0.83em;
  margin-block-end: 0.83em;
}
.kfp-start-page h2 {
  font-size: 18px;
  margin-block-start: 1em;
  margin-block-end: 1em;
}
.kfp-start-page h3 {
  font-size: 16px;
  margin-block-start: 1em;
  margin-block-end: 1em;
}
`);

const OPTIONS = {
  overrides: { a: { component: AutoLink } },
};

export class GettingStarted extends Page<{}, { links: string[] }> {
  public state = {
    links: ['', '', '', '', ''].map(getPipelineLink),
  };

  public getInitialToolbarState(): ToolbarProps {
    const buttons = new Buttons(this.props, this.refresh.bind(this));
    return {
      actions: buttons.getToolbarActionMap(),
      breadcrumbs: [],
      pageTitle: 'Getting Started',
    };
  }

  public async componentDidMount() {
    const ids = await Promise.all(
      DEMO_PIPELINES.map(name =>
        Apis.pipelineServiceApi
          .listPipelines(undefined, 10, undefined, createAndEncodeFilter(name))
          .then(pipelineList => {
            const pipelines = pipelineList.pipelines;
            if (pipelines?.length !== 1) {
              // This should be accurate, do not accept ambiguous results.
              return '';
            }
            return pipelines[0].id || '';
          })
          .catch(() => ''),
      ),
    );
    this.setState({ links: ids.map(getPipelineLink) });
  }

  public async refresh() {
    this.componentDidMount();
  }

  public render(): JSX.Element {
    return (
      <div className={classes(commonCss.page, padding(20, 'lr'), 'kfp-start-page')}>
        <Markdown options={OPTIONS}>
          {PAGE_CONTENT_MD({
            control: this.state.links[DEMO_PIPELINES_ID_MAP.control],
            data: this.state.links[DEMO_PIPELINES_ID_MAP.data],
            tfx: this.state.links[DEMO_PIPELINES_ID_MAP.tfx],
            xgboost: this.state.links[DEMO_PIPELINES_ID_MAP.xgboost],
            v2python: this.state.links[DEMO_PIPELINES_ID_MAP.v2python],
          })}
        </Markdown>
      </div>
    );
  }
}

function getPipelineLink(id: string) {
  if (!id) {
    return '#/pipelines';
  }
  return `#${RoutePageFactory.pipelineDetails(id)}`;
}

function createAndEncodeFilter(filterString: string): string {
  const filter: ApiFilter = {
    predicates: [
      {
        key: 'name',
        op: PredicateOp.EQUALS,
        string_value: filterString,
      },
    ],
  };
  return encodeURIComponent(JSON.stringify(filter));
}
