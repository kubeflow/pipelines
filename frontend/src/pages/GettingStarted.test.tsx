/**
 * Copyright 2021 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { GettingStarted } from './GettingStarted';
import TestUtils, { diffHTML } from '../TestUtils';
import { render } from '@testing-library/react';
import { PageProps } from './Page';
import { Apis } from '../lib/Apis';
import { ApiListPipelinesResponse } from '../apis/pipeline/api';

const PATH_BACKEND_CONFIG = '../../../backend/src/apiserver/config/sample_config.json';
const PATH_FRONTEND_CONFIG = '../config/sample_config_from_backend.json';
describe(`${PATH_FRONTEND_CONFIG}`, () => {
  it(`should be in sync with ${PATH_BACKEND_CONFIG}, if not please run "npm run sync-backend-sample-config" to update.`, () => {
    const configBackend = require(PATH_BACKEND_CONFIG);
    const configFrontend = require(PATH_FRONTEND_CONFIG);
    expect(configFrontend).toEqual(configBackend.map((sample: any) => sample.name));
  });
});

describe('GettingStarted page', () => {
  const updateBannerSpy = jest.fn();
  const updateToolbarSpy = jest.fn();
  const historyPushSpy = jest.fn();
  const pipelineListSpy = jest.spyOn(Apis.pipelineServiceApi, 'listPipelines');

  function generateProps(): PageProps {
    return TestUtils.generatePageProps(
      GettingStarted,
      {} as any,
      {} as any,
      historyPushSpy,
      updateBannerSpy,
      null,
      updateToolbarSpy,
      null,
    );
  }

  beforeEach(() => {
    jest.resetAllMocks();
    const empty: ApiListPipelinesResponse = {
      pipelines: [],
      total_size: 0,
    };
    pipelineListSpy.mockImplementation(() => Promise.resolve(empty));
  });

  it('initially renders documentation', () => {
    const { container } = render(<GettingStarted {...generateProps()} />);
    expect(container).toMatchSnapshot();
  });

  it('renders documentation with pipeline deep link after querying demo pipelines', async () => {
    let count = 0;
    pipelineListSpy.mockImplementation(() => {
      ++count;
      const response: ApiListPipelinesResponse = {
        pipelines: [{ id: `pipeline-id-${count}` }],
      };
      return Promise.resolve(response);
    });
    const { container } = render(<GettingStarted {...generateProps()} />);
    const base = container.innerHTML;
    await TestUtils.flushPromises();
    expect(pipelineListSpy.mock.calls).toMatchSnapshot();
    expect(diffHTML({ base, update: container.innerHTML })).toMatchInlineSnapshot(`
      Snapshot Diff:
      - Expected
      + Received

      @@ --- --- @@
            <h2 id="demonstrations-and-tutorials">Demonstrations and Tutorials</h2>
            <p>This section contains demo and tutorial pipelines.</p>
            <p><strong>Demos</strong> - Try an end-to-end demonstration pipeline.</p>
            <ul>
              <li>
      -         <a href="#/pipelines" class="link">TFX pipeline demo with Estimator</a>
      +         <a href="#/pipelines/details/pipeline-id-2?" class="link"
      +           >TFX pipeline demo with Estimator</a
      +         >
                <ul>
                  <li>
                    Classification pipeline with model analysis, based on a public
                    BigQuery dataset of taxicab trips.
                    <a
      @@ --- --- @@
                    >
                  </li>
                </ul>
              </li>
              <li>
      -         <a href="#/pipelines" class="link">XGBoost Pipeline demo</a>
      +         <a href="#/pipelines/details/pipeline-id-1?" class="link"
      +           >XGBoost Pipeline demo</a
      +         >
                <ul>
                  <li>
                    An example of end-to-end iterative XGBoost model training.
                    <a
                      href="https://github.com/kubeflow/pipelines/tree/master/samples/core/train_until_good"
      @@ --- --- @@
              <strong>Tutorials</strong> - Learn pipeline concepts by following a
              tutorial.
            </p>
            <ul>
              <li>
      -         <a href="#/pipelines" class="link">Data passing in Python components</a>
      +         <a href="#/pipelines/details/pipeline-id-3?" class="link"
      +           >Data passing in Python components</a
      +         >
                <ul>
                  <li>
                    Shows how to pass data between Python components.
                    <a
                      href="https://github.com/kubeflow/pipelines/tree/master/samples/tutorials/Data%20passing%20in%20python%20components"
      @@ --- --- @@
                    >
                  </li>
                </ul>
              </li>
              <li>
      -         <a href="#/pipelines" class="link">DSL - Control structures</a>
      +         <a href="#/pipelines/details/pipeline-id-4?" class="link"
      +           >DSL - Control structures</a
      +         >
                <ul>
                  <li>
                    Shows how to use conditional execution and exit handlers.
                    <a
                      href="https://github.com/kubeflow/pipelines/tree/master/samples/tutorials/DSL%20-%20Control%20structures"
    `);
  });

  it('fallbacks to show pipeline list page if request failed', async () => {
    let count = 0;
    pipelineListSpy.mockImplementation(
      (): Promise<ApiListPipelinesResponse> => {
        ++count;
        if (count === 1) {
          return Promise.reject(new Error('Mocked error'));
        } else if (count === 2) {
          // incomplete data
          return Promise.resolve({});
        } else if (count === 3) {
          // empty data
          return Promise.resolve({ pipelines: [], total_size: 0 });
        }
        return Promise.resolve({
          pipelines: [{ id: `pipeline-id-${count}` }],
          total_size: 1,
        });
      },
    );
    const { container } = render(<GettingStarted {...generateProps()} />);
    const base = container.innerHTML;
    await TestUtils.flushPromises();
    expect(diffHTML({ base, update: container.innerHTML })).toMatchInlineSnapshot(`
      Snapshot Diff:
      - Expected
      + Received

      @@ --- --- @@
                    >
                  </li>
                </ul>
              </li>
              <li>
      -         <a href="#/pipelines" class="link">DSL - Control structures</a>
      +         <a href="#/pipelines/details/pipeline-id-4?" class="link"
      +           >DSL - Control structures</a
      +         >
                <ul>
                  <li>
                    Shows how to use conditional execution and exit handlers.
                    <a
                      href="https://github.com/kubeflow/pipelines/tree/master/samples/tutorials/DSL%20-%20Control%20structures"
    `);
  });
});
