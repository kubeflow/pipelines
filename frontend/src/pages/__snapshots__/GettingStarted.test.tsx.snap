// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GettingStarted page initially renders documentation 1`] = `
<div>
  <div
    class="page kfp-start-page"
  >
    <div>
      <br />
      <h2
        id="build-your-own-pipeline-with"
      >
        Build your own pipeline with
      </h2>
      <ul>
        <li>
          Kubeflow Pipelines 
          <a
            class="link"
            href="https://www.kubeflow.org/docs/pipelines/sdk/"
            rel="noopener"
            target="_blank"
          >
            SDK
          </a>
          

          <ul>
            <li>
              Try out 
              <a
                class="link"
                href="https://www.kubeflow.org/docs/components/pipelines/sdk/v2/v2-compatibility/"
                rel="noopener"
                target="_blank"
              >
                the new Kubeflow Pipelines SDK v2 (Beta)
              </a>
            </li>
          </ul>
        </li>
        <li>
          TensorFlow Extended (TFX) 
          <a
            class="link"
            href="https://www.tensorflow.org/tfx/guide"
            rel="noopener"
            target="_blank"
          >
            SDK
          </a>
           with end-to-end ML Pipeline Template (
          <a
            class="link"
            href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?q=download_url%3Dhttps%253A%252F%252Fraw.githubusercontent.com%252Ftensorflow%252Ftfx%252Fmaster%252Fdocs%252Ftutorials%252Ftfx%252Ftemplate.ipynb"
            rel="noopener"
            target="_blank"
          >
            Open a Vertex AI Workbench on Google Cloud
          </a>
          )
        </li>
      </ul>
      <br />
      <h2
        id="demonstrations-and-tutorials"
      >
        Demonstrations and Tutorials
      </h2>
      <p>
        This section contains demo and tutorial pipelines.
      </p>
      <p>
        <strong>
          Demos
        </strong>
         - Try an end-to-end demonstration pipeline.
      </p>
      <ul>
        <li>
          <a
            class="link"
            href="#/pipelines"
          >
            TFX pipeline demo with Estimator
          </a>
           
          <ul>
            <li>
              Classification pipeline with model analysis, based on a public BigQuery dataset of taxicab trips. 
              <a
                class="link"
                href="https://github.com/kubeflow/pipelines/tree/master/samples/core/parameterized_tfx_oss"
                rel="noopener"
                target="_blank"
              >
                source code
              </a>
            </li>
          </ul>
        </li>
        <li>
          <a
            class="link"
            href="#/pipelines"
          >
            XGBoost Pipeline demo
          </a>
           
          <ul>
            <li>
              An example of end-to-end iterative XGBoost model training. 
              <a
                class="link"
                href="https://github.com/kubeflow/pipelines/tree/master/samples/core/train_until_good"
                rel="noopener"
                target="_blank"
              >
                source code
              </a>
            </li>
          </ul>
        </li>
      </ul>
      <br />
      <p>
        <strong>
          Tutorials
        </strong>
         - Learn pipeline concepts by following a tutorial.
      </p>
      <ul>
        <li>
          <a
            class="link"
            href="#/pipelines"
          >
            Data passing in Python components
          </a>
           
          <ul>
            <li>
              Shows how to pass data between Python components. 
              <a
                class="link"
                href="https://github.com/kubeflow/pipelines/tree/master/samples/tutorials/Data%20passing%20in%20python%20components"
                rel="noopener"
                target="_blank"
              >
                source code
              </a>
            </li>
          </ul>
        </li>
        <li>
          <a
            class="link"
            href="#/pipelines"
          >
            DSL - Control structures
          </a>
           
          <ul>
            <li>
              Shows how to use conditional execution and exit handlers. 
              <a
                class="link"
                href="https://github.com/kubeflow/pipelines/tree/master/samples/tutorials/DSL%20-%20Control%20structures"
                rel="noopener"
                target="_blank"
              >
                source code
              </a>
            </li>
          </ul>
        </li>
      </ul>
      <p>
        Want to learn more? 
        <a
          class="link"
          href="https://www.kubeflow.org/docs/pipelines/tutorials/"
          rel="noopener"
          target="_blank"
        >
          Learn from sample and tutorial pipelines.
        </a>
      </p>
    </div>
  </div>
</div>
`;

exports[`GettingStarted page renders documentation with pipeline deep link after querying demo pipelines 1`] = `
Array [
  Array [
    undefined,
    10,
    undefined,
    "%7B%22predicates%22%3A%5B%7B%22key%22%3A%22name%22%2C%22op%22%3A%22EQUALS%22%2C%22string_value%22%3A%22%5BDemo%5D%20XGBoost%20-%20Iterative%20model%20training%22%7D%5D%7D",
  ],
  Array [
    undefined,
    10,
    undefined,
    "%7B%22predicates%22%3A%5B%7B%22key%22%3A%22name%22%2C%22op%22%3A%22EQUALS%22%2C%22string_value%22%3A%22%5BDemo%5D%20TFX%20-%20Taxi%20tip%20prediction%20model%20trainer%22%7D%5D%7D",
  ],
  Array [
    undefined,
    10,
    undefined,
    "%7B%22predicates%22%3A%5B%7B%22key%22%3A%22name%22%2C%22op%22%3A%22EQUALS%22%2C%22string_value%22%3A%22%5BTutorial%5D%20Data%20passing%20in%20python%20components%22%7D%5D%7D",
  ],
  Array [
    undefined,
    10,
    undefined,
    "%7B%22predicates%22%3A%5B%7B%22key%22%3A%22name%22%2C%22op%22%3A%22EQUALS%22%2C%22string_value%22%3A%22%5BTutorial%5D%20DSL%20-%20Control%20structures%22%7D%5D%7D",
  ],
]
`;
